tests/test_requests.py

```diff
@@ -623,6 +623,21 @@ def test_DIGESTAUTH_QUOTES_QOP_VALUE(self, httpbin):
         r = requests.get(url, auth=auth)
         assert '"auth"' in r.request.headers['Authorization']
 
+    def test_DIGESTAUTH_DOESNT_RESEND(self, httpbin):
+        """If HTTPDigestAuth fails a 401 challenge, it won't try again"""
+        s = requests.Session()
+        auth = requests.auth.HTTPDigestAuth('user', 'pass')
+
+        r = requests.Request('GET', httpbin('digest-auth/auth/user/pass'), auth=auth)
+        p = r.prepare()
+
+        # simulate previously failed 401 so auth fails
+        auth._thread_local.num_401_calls = 2
```

Comment:
So this was originally written for specifically testing `handle_401` as a method, now it's morphed into leaning towards testing a behaviour. I'll rewrite this as a socket test to make sure HTTPDigestAuth functions correctly rather than focusing on the hook.

I do think removing this will make it a bit harder to pinpoint the problem in a future failure though. I understand the hesitance to not test a variable that may be ephemeral, but only having the end-to-end tests kind of hides the smoking gun. There have been previous instances of HTTPDigestAuth failing specifically because `num_401_calls` was modified incorrectly and there were no tests to catch it.

`num_401_calls` is a well defined attribute that's been the core of how HTTPDigestAuth functions over the wire for 4 years.  We know that it's an unconventional boolean that will be either 1 and 2 at any point after it's been invoked. Unless we're planning to rewrite HTTPDigestAuth in the near future, wouldn't we want more granularity in the tests where there have been previous failures?

Sorry if I'm coming across as pig headed over such an minor issue, I'm just trying to understand how you'd solve this.